bigtable-dataflow-parent/bigtable-beam-import/README.md

```diff
@@ -1,44 +1,220 @@
-Cloud Dataflow.
+This folder contains tools to support importing and exporting HBase data to
+Google Cloud Bigtable using Cloud Dataflow.
 
-## Instructions
-[//]: # ({x-version-update-start:bigtable-dataflow-parent:released})
-Download [the import/export jar](http://search.maven.org/remotecontent?filepath=com/google/cloud/bigtable/bigtable-beam-import/1.14.1/bigtable-beam-import-1.14.1-shaded.jar), which is an aggregation of all required jars.
+## Setup 
 
-Please pay attention to the Cluster CPU usage and adjust the number of Dataflow workers accordingly.
+To use the tools in this folder, you can download them from the maven repository, or
+you can build them using Maven. 
+
+
+[//]: # ({x-version-update-start:bigtable-dataflow-parent:released})
+### Download the jars
+Download [the import/export jars](http://search.maven.org/remotecontent?filepath=com/google/cloud/bigtable/bigtable-beam-import/1.14.1/bigtable-beam-import-1.14.1-shaded.jar), which is an aggregation of all required jars.
 
-## Export
+### Build the jars yourself
 
-On the command line:
+Go to the top level directory and build the repo
+then return to this sub directory.
 
 ```
-java -jar bigtable-beam-import-1.14.1-shaded.jar export \
-    --runner=dataflow \
-    --project=[your_project_id] \
-    --bigtableInstanceId=[your_instance_id] \
-    --bigtableTableId=[your_table_id] \
-    --destinationPath=gs://[bucket_name]/[export_directory]/ \
-    --tempLocation=gs://[bucket_name]/[temp_work_directory]/ \
-    --maxNumWorkers=[10x number of nodes] \
+cd ../../
+mvn clean install -DskipTests=true
+cd bigtable-dataflow-parent/bigtable-beam-import
 ```
 
-## Import
+***
+#Tools
 
-Create the table in your cluster.
+## Data export pipeline
+
+You can export data into a snapshot or into sequence files. If you're migrating
+your data from HBase to Bigtable using snapshots is the preferred method. 
+
+### Exporting snapshots from HBase
+
+Perform these steps from Unix shell on HBase master node.
+
+1. Set the environment variables
+    ```
+    TABLE_NAME=your-table-name
+    SNAPSHOT_NAME=your-snapshot-name 
+    SNAPSHOT_EXPORT_PATH=/hbase-migration-snap
+    BUCKET_NAME="gs://bucket-name"
+    ```
+1. Take the snapshot
+    ```
+    echo "snapshot '$TABLE_NAME', '$SNAPSHOT_NAME'" | hbase shell -n
+    ```
+
+1. Export the snapshot   
+    ```
+    hbase org.apache.hadoop.hbase.snapshot.ExportSnapshot -snapshot $SNAPSHOT_NAME \
+        -copy-to $SNAPSHOT_EXPORT_PATH/data -mappers 16
+    ```
+1. Create hashes for the table to be used during the data validation step.
+   ```
+   hbase org.apache.hadoop.hbase.mapreduce.HashTable --batchsize=10 --numhashfiles=10 \
+   $TABLE_NAME $SNAPSHOT_EXPORT_PATH/hashtable
+   ```
+1. Export the data and hashes to a GCS bucket using [gsutil](https://cloud.google.com/storage/docs/gsutil).
+   ```
+   hadoop fs -copyToLocal $SNAPSHOT_EXPORT_PATH /tmp/
+   gsutil -m cp -r /tmp$SNAPSHOT_EXPORT_PATH $BUCKET_NAME
+   ``` 
```

Comment:
* `ExportSnapshot` is a distributed copy tool. We would copy the snapshot directly to gcs in this step vs. copyToLocal & use gsutil. 
* The output of `HashTable` can write directly to gcs as well. 
* In order to write to gcs from an on-prem hadoop cluster, the [GCS connector is required](https://cloud.google.com/dataproc/docs/concepts/connectors/install-storage-connector) for the application to write to GCS.

Reply:
So instead of doing `-copy-to $SNAPSHOT_EXPORT_PATH/data` I can say `-copy-to gs://my-bucket/data` ?

Reply:
yup, exactly. and for HashTable, `$SNAPSHOT_EXPORT_PATH/hashtable` can be updated to `gs://my-bucket/hashtable`